File: 2023/day_01.exs
#!/usr/bin/env elixir

defmodule Day01 do
  @moduledoc """
  Day 1 - Trebuchet?!

  Part 1:
    Recover calibration values from a document by keeping the first and last
    digit found in each line. Sum all the two-digit numbers.

  Part 2:
    Extends part 1 by also including the overlapping spelled-out digits as
    possible valid digits.
  """

  def parse(""), do: []
  def parse(<<c>>   <> rest) when c in ?0..?9, do: [c - ?0 | parse(rest)]
  def parse("zero"  <> rest), do: [0 | parse("o" <> rest)]
  def parse("one"   <> rest), do: [1 | parse("e" <> rest)]
  def parse("two"   <> rest), do: [2 | parse("o" <> rest)]
  def parse("three" <> rest), do: [3 | parse("e" <> rest)]
  def parse("four"  <> rest), do: [4 | parse("r" <> rest)]
  def parse("five"  <> rest), do: [5 | parse("e" <> rest)]
  def parse("six"   <> rest), do: [6 | parse("x" <> rest)]
  def parse("seven" <> rest), do: [7 | parse("n" <> rest)]
  def parse("eight" <> rest), do: [8 | parse("t" <> rest)]
  def parse("nine"  <> rest), do: [9 | parse("e" <> rest)]
  def parse(<<_>>   <> rest), do: parse(rest)

  def solve(str) do
    String.split(str)
    |> Enum.map(&Day01.parse/1)
    |> Enum.map(fn digits -> List.first(digits) * 10 + List.last(digits) end)
    |> Enum.sum()
  end
end

part_1 = File.read!("./inputs/example_01_1.txt")
part_2 = File.read!("./inputs/example_01_2.txt")

IO.inspect(Day01.solve(part_1))
IO.inspect(Day01.solve(part_2))

File: 2023/ruby/day_01.rb
#!/usr/bin/env ruby

require_relative '../../aocday'

class Day01 < AoCDay
  WORD_TO_INTEGER = {
    'zero' => 0, 'one' => 1, 'two' => 2, 'three' => 3,
    'four' => 4, 'five' => 5, 'six' => 6, 'seven' => 7,
    'eight' => 8, 'nine' => 9,
  }

  def setup(data)
    @lines = data.split("\n")
  end

  def extract_calibration_value_v1(line)
    line
      .gsub(/\D/, '')
      .chars.values_at(0, -1)
      .join.to_i
  end

  def extract_calibration_value_v2(line)
    line
      .scan(/(?=(zero|one|two|three|four|five|six|seven|eight|nine|\d))/)
      .map { |match| to_word_i(match.first) }
      .values_at(0, -1)
      .join.to_i
  end

  def part1
    @lines.map { |line| extract_calibration_value_v1(line) }.sum
  end

  def part2
    @lines.map { |line| extract_calibration_value_v2(line) }.sum
  end

  private

  def to_word_i(string)
    WORD_TO_INTEGER[string] || string.to_i
  end
end
